src/Qir/Runtime/lib/QIR/QubitManager.cpp

```diff
@@ -0,0 +1,508 @@
+    DeleteQubitObject(qubit);
+}
+
+void CQubitManager::Release(Qubit* qubitsToRelease, int32_t qubitCountToRelease) {
+    if (qubitCountToRelease == 0)
+    {
+        return;
+    }
+    FailIf(qubitCountToRelease < 0, "Cannot release negative number of qubits.");
+    FailIf(qubitsToRelease == nullptr, "No array provided with qubits to be released.");
+
+    for (int32_t i = 0; i < qubitCountToRelease; i++)
+    {
+        Release(qubitsToRelease[i]);
+        qubitsToRelease[i] = nullptr;
+    }
+}
+
+Qubit CQubitManager::Borrow()
+{
+    // We don't support true borrowing/returning at the moment.
+    return Allocate();
+}
+
+void CQubitManager::Borrow(Qubit* qubitsToBorrow, int32_t qubitCountToBorrow)
+{
+    // We don't support true borrowing/returning at the moment.
+    return Allocate(qubitsToBorrow, qubitCountToBorrow);
+}
+
+void CQubitManager::Return(Qubit qubit)
+{
+    // We don't support true borrowing/returning at the moment.
+    Release(qubit);
+}
+
+void CQubitManager::Return(Qubit* qubitsToReturn, int32_t qubitCountToReturn)
+{
+    // We don't support true borrowing/returning at the moment.
+    Release(qubitsToReturn, qubitCountToReturn);
+}
+
+void CQubitManager::Disable(Qubit qubit)
+{
+    FailIf(!IsValidQubit(qubit), "Qubit is not valid.");
+    QubitIdType id = QubitToId(qubit);
+
+    // We can only disable explicitly allocated qubits that were not borrowed.
+    FailIf(!IsExplicitlyAllocated(id), "Cannot disable qubit that is not explicitly allocated.");
+    sharedQubitStatusArray[id] = DisabledMarker;
+
+    disabledQubitCount++;
+    FailIf(disabledQubitCount <= 0, "Incorrect disabled qubit count.");
+    allocatedQubitCount--;
+    FailIf(allocatedQubitCount < 0, "Incorrect allocated qubit count.");
+}
+
+void CQubitManager::Disable(Qubit* qubitsToDisable, int32_t qubitCountToDisable)
+{
+    if (qubitCountToDisable == 0)
+    {
+        return;
+    }
+    FailIf(qubitCountToDisable < 0, "Cannot disable negative number of qubits.");
+    FailIf(qubitsToDisable == nullptr, "No array provided with qubits to be disabled.");
+
+    for (int32_t i = 0; i < qubitCountToDisable; i++)
+    {
+        Disable(qubitsToDisable[i]);
+    }
+}
+
+bool CQubitManager::IsValidQubit(Qubit qubit) const
+{
+    QubitIdType id = QubitToId(qubit);
+    if (id >= qubitCapacity)
+    {
+        return false;
+    }
+    if (id < 0)
+    {
+        return false;
+    }
+    return true;
+}
+
+bool CQubitManager::IsDisabledQubit(Qubit qubit) const
+{
+    return IsValidQubit(qubit) && IsDisabled(QubitToId(qubit));
+}
+
+bool CQubitManager::IsFreeQubit(Qubit qubit) const
+{
+    return IsValidQubit(qubit) && IsFree(QubitToId(qubit));
+}
+
+CQubitManager::QubitIdType CQubitManager::GetQubitId(Qubit qubit) const
+{
+    FailIf(!IsValidQubit(qubit), "Not a valid qubit.");
+    return QubitToId(qubit);
+}
+
+
+Qubit CQubitManager::CreateQubitObject(QubitIdType id)
+{
+    FailIf(id < 0 || id > INTPTR_MAX, "Qubit id is out of range.");
```

Comment:
`INTPTR_MAX` is a 64-bit value on a 64-bit platform, but `id` is 32-bit.
```suggestion
    FailIf(id < 0 || id >= MaximumQubitCapacity, "Qubit id is out of range.");
```

Reply:
Or maybe even `qubitCapacity` instead of `MaximumQubitCapacity`.

Reply:
The intention of the check is to make sure the qubit id fits into intptr_t to which it is casted on the next line. This is a good candidate for a debug only check or even static assert.

Reply:
On a second though, this should stay as is, at least for now. Regardless of how wide QubitIdType and intptr_t types are, this ensures that static cast that follows works just fine as it checks that the value will fit.

Reply:
FYI:
```
pwsh bootstrap.ps1

. . .
[14/35] Building CXX object lib/QIR/CMakeFiles/qir-rt-support-obj.dir/QubitManager.cpp.obj
../../lib/QIR/QubitManager.cpp:395:25: warning: result of comparison of constant 9223372036854775807 with expression of type 'Microsoft::Quantum::CQubitManager::QubitIdType' (aka 'int') is always false [-Wtautological-constant-out-of-range-compare]
    FailIf(id < 0 || id > std::numeric_limits<intptr_t>::max(), "Qubit id is out of range.");
                     ~~ ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
```
If you get a chance soon please fix.
(we are using clang-11, but locally I have clang 12 and it reports. So you might not be able to repro with clang-11)

Reply:
FYI: Has been changed in [my PR](https://github.com/microsoft/qsharp-runtime/pull/718#discussion_r647046753). Feel free to change/correct in the subsequent PRs.